sprint_10

SelyukSergey opened the pull request.

SelyukSergey committed:
Added secret keys

SelyukSergey committed:
Реализация авторизации через OAuth2.0: добавление констант, WebView, SplashViewController, OAuth2Service, логирование ошибок, финализация дизайна

SelyukSergey committed:
.

Kirchberg commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,109 @@
+import UIKit
+import WebKit
+
+fileprivate let UnsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Лучше ограничивать область видимости констант. Если константа используется только внутри класса, имеет смысл объявить её как private static (или просто private, если контекст позволяет) внутри класса - тогда она будет именоваться в lowerCamelCase.

Kirchberg commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,109 @@
+
+final class WebViewViewController: UIViewController {
+    
+    // MARK: - IBOutlets
```
Что касается комментариев с MARK, то рекомендуется оставлять пробел после знака «:» и использовать единообразный стиль, например:
```swift
// MARK: - IBOutlets
```
Также удобно отделять секции кода пустыми строками, чтобы визуально разграничивать разные функциональные блоки - это улучшает навигацию по файлу и делает код более структурированным.

Kirchberg commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,46 @@
+    weak var delegate: AuthViewControllerDelegate?
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == ShowWebViewSegueIdentifier {
```
Можно использовать в данном случае конструкцию guard:
```swift
guard segue.identifier == ShowWebViewSegueIdentifier else {
    super.prepare(for: segue, sender: sender)
    return
}
...
```

Kirchberg commented on ImageFeed/Constants.swift:
⚠️ Необходимо поправить файл `Constants.swift` и объявить внутри него `enum Constants` внутри которого находятся все статические константы начинаются с маленькой буквы, например, `static let accessKey = ""`, поправить необходимо в соответствии с условие задания 6.1.

Kirchberg commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,86 @@
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        if let token = oauth2TokenStorage.token {
```
В данном случае нам не нужна константа, поэтому можно проверить ее наличие таким образом:
```swift
if oauth2TokenStorage.token != nil {
    ...
} else {
    ...
}
```

В данном случае даже можно использовать конструкцию guard

Kirchberg commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,81 @@
+    private var authToken: String?
+    
+    // MARK: - Request Creation
+    func makeOAuthTokenRequest(code: String) -> URLRequest {
+        let baseURL = URL(string: "https://unsplash.com")!
+        let url = URL(
+            string: "/oauth/token"
+            + "?client_id=\(AccessKey)"
+            + "&&client_secret=\(SecretKey)"
+            + "&&redirect_uri=\(RedirectURI)"
+            + "&&code=\(code)"
+            + "&&grant_type=authorization_code",
+            relativeTo: baseURL
+        )!
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        return request
+    }
```
⚠️ В проекте используется `implicit unwrap` в методе `makeOAuthTokenRequest`, необходимо сделать безопасное извлечение значений опционалов и залогировать ошибочные кейсы в соответствии с условием пунктов задания 4 и 12.1.

Kirchberg commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,81 @@
+    // MARK: - Fetch OAuth Token
+    func fetchOAuthToken(_ code: String, completion: @escaping (Result<String, any Error>) -> Void) {
+        let request = makeOAuthTokenRequest(code: code)
+        
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+            if let error = error {
+                DispatchQueue.main.async {
+                    completion(.failure(error))
+                }
+                return
+            }
+            
+            guard let data = data else {
+                DispatchQueue.main.async {
+                    completion(.failure(NSError(domain: "NoData", code: -1, userInfo: nil)))
+                }
+                return
+            }
+            
+            do {
+                let decoder = JSONDecoder()
+                decoder.keyDecodingStrategy = .convertFromSnakeCase
+                let OAuthTokenResponseBody = try decoder.decode(OAuthTokenResponseBody.self, from: data)
+                self.authToken = OAuthTokenResponseBody.accessToken
+                
+                // save token to storage
+                OAuth2TokenStorage.shared.token = OAuthTokenResponseBody.accessToken
+                
+                DispatchQueue.main.async {
+                    completion(.success(OAuthTokenResponseBody.accessToken))
+                }
+            } catch {
+                DispatchQueue.main.async {
+                    completion(.failure(error))
+                }
+            }
+        }
+        task.resume()
+    }
+}
```
⚠️ Не залогированы ошибки при получении сетевых ошибок при работе метода `URLSession.shared.dataTask`, а также ошибки, которую может выкинуть декодер при получении `OAuthTokenResponseBody`, необходимо поправить в соответствии с пунктами задания 12.2 и 12.4

Kirchberg commented on ImageFeed/URLSession+data.swift:
```diff
@@ -0,0 +1,36 @@
+            }
+        }
+        
+        let task = dataTask(with: request, completionHandler: { data, response, error in
+            if let data = data, let response = response, let statusCode = (response as? HTTPURLResponse)?.statusCode {
+                if 200 ..< 300 ~= statusCode {
+                    fulfillCompletionOnTheMainThread(.success(data))
+                } else {
+                    fulfillCompletionOnTheMainThread(.failure(NetworkError.httpStatusCode(statusCode)))
+                }
+            } else if let error = error {
+                fulfillCompletionOnTheMainThread(.failure(NetworkError.urlRequestError(error)))
+            } else {
+                fulfillCompletionOnTheMainThread(.failure(NetworkError.urlSessionError)) 
+            }
+        })
```
⚠️ Не залогированы ошибки, которые вернул сервис Unsplash (с кодом ответа >=300), необходимо поправить в соответствии с пунктом задания 12.3

Kirchberg reviewed (changes requested):
⚠️ Критические замечания (это нужно исправить обязательно):

1. Необходимо поправить файл `Constants.swift` и объявить внутри него `enum Constants` внутри которого находятся все статические константы начинаются с маленькой буквы, например, `static let accessKey = ""`, поправить необходимо в соответствии с условие задания 6.1.
2. В проекте используется `implicit unwrap` в методе `makeOAuthTokenRequest`, необходимо сделать безопасное извлечение значений опционалов и залогировать ошибочные кейсы в соответствии с условием пунктов задания 4 и 12.1.
3. Не залогированы ошибки при получении сетевых ошибок при работе метода `URLSession.shared.dataTask`, а также ошибки, которую может выкинуть декодер при получении `OAuthTokenResponseBody`, необходимо поправить в соответствии с пунктами задания 12.2 и 12.4.
4. Не залогированы ошибки, которые вернул сервис Unsplash (с кодом ответа >=300), необходимо поправить в соответствии с пунктом задания 12.3.

Без исправлений этих пунктов, я не смогу принять у тебя работу. Если возникнут вопросы, обязательно задавай их Наставнику!

SelyukSergey committed:
Добавлено логирование ошибок в консоль
Константы описаны внутри перечисления(enum) Constants статическими константами
Исправлены прочие недочеты

Kirchberg reviewed (approved):
Не забудь вмерджить пул-реквест в основную ветку.
Успехов в следующем задании!
